sway-fmt-v2/src/config/manifest.rs

```diff
@@ -0,0 +1,260 @@
+            structures: Structures::default(),
+            comments: Comments::default(),
+        }
+    }
+    /// Given an optional path to a `swayfmt.toml`, read it and construct a `FormatConfig`.
+    /// If settings are omitted, those fields will be set to default. If `None` is provided,
+    /// the default config will be applied. If a `swayfmt.toml` exists but is empty, the default
+    /// config will be applied.
+    ///
+    /// At present, this will only return a warning if it catches unusable fields.
+    /// Upon completion, this should give errors/warnings for incorrect input fields as well.
+    ///
+    pub fn from_dir_or_default(config_path: Option<&Path>) -> Result<Self> {
+        let config = OptionFormatConfig::from_dir_or_default(config_path)?;
+        Ok(config)
+    }
+}
+
+impl OptionFormatConfig {
+    /// Given an optional path to a `swayfmt.toml`, read it and construct a `FormatConfig`.
+    /// If settings are omitted, those fields will be set to default. If `None` is provided,
+    /// the default config will be applied. If a `swayfmt.toml` exists but is empty, the default
+    /// config will be applied.
+    ///
+    /// At present, this will only return a warning if it catches unusable fields.
+    /// Upon completion, this should give errors/warnings for incorrect input fields as well.
+    ///
+    pub fn from_dir_or_default(config_path: Option<&Path>) -> Result<FormatConfig> {
+        match config_path {
+            Some(starter_path) => {
+                if let Some(path) = find_parent_dir_with_file(starter_path, SWAY_FORMAT_FILE_NAME) {
+                    let config_str = std::fs::read_to_string(path)
+                        .map_err(|e| anyhow!("failed to read config at {:?}: {}", path, e))?;
+                    // save some time if the file is empty
+                    if !config_str.is_empty() {
+                        let toml_de = &mut toml::de::Deserializer::new(&config_str);
+                        let user_settings: Self = serde_ignored::deserialize(toml_de, |field| {
+                            let warning =
+                                format!("  WARNING! found unusable configuration: {}", field);
+                            println_yellow_err(&warning);
+                        })
+                        .map_err(|e| anyhow!("failed to parse config: {}.", e))?;
+
+                        Ok(Self::apply_user_settings(user_settings))
+                    } else {
+                        Ok(FormatConfig::default())
+                    }
+                } else {
+                    Ok(FormatConfig::default())
+                }
+            }
+            None => Ok(FormatConfig::default()),
+        }
+    }
+    /// Check the user's settings, and replace the values of the default formatter
+    /// with those of the user's if they exist.
+    ///
+    pub fn apply_user_settings(user_settings: OptionFormatConfig) -> FormatConfig {
```

Comment:
@JoshuaBatty if there's an easier/nicer looking way of doing this, I'm all ears

Reply:
@mitchmindtree also feel free to weigh in, in general

Reply:
I think a nicer way to do this might be to implement this as a constructor on `FormatConfig`, something like:

```rust
impl FormatConfig {
    /// Construct the set of configuration to be used from the given set of options.
    pub fn from_opts(opts: &OptionFormatConfig) -> Self {
        Self {
            whitespace: opts.whitespace.as_ref().map(Whitespace::from_opts).unwrap_or_default(),
            imports: opts.imports.as_ref().map(Imports::from_opts).unwrap_or_default(),
            // etc.
        }
    }
}
```

Then also have something for each of the subsets of options (i.e. `Whitespace`, `Imports`), like:

```rust
impl Whitespace {
    pub const DEFAULT_MAX_WIDTH: usize = 100;
    pub const DEFAULT_HARD_TABS: bool = false;
    // the rest of the defaults...

    pub fn from_opts(opts: &WhitespaceOptions) -> Self {
        Self {
            max_width: opts.max_width.unwrap_or(Whitespace::DEFAULT_MAX_WIDTH),
            hard_tabs: opts.hard_tabs.unwrap_or(Whitespace::DEFAULT_HARD_TABS),
            // etc
        }
    }
}
```

Also I noticed the top-level options are called `OptionFormatConfig`, but the subsets are named `*Options` (i.e. `WhitespaceOptions`, `ImportsOptions`, etc). Maybe we should rename the top-level type to be consistent, e.g. `FormatConfigOptions`.

Also, we might be able to omit the `Format` word from these types altogether as it's kind of implied by being in the formatter crate, e.g. `sway_fmt::Config`. Maybe `Config` and `ConfigOptions` would be fine names for these?

Reply:
@mitchmindtree thanks so much, I'll get this sorted tomorrow morning!

Reply:
Is it necessary to have constants for the boolean defaults?

Reply:
I think it's generally good practise to publicly expose defaults in a consistent manner, however I think an exception can be made in the case that the default value is equivalent to the `Default` implementation of its type (as is the case for `hard_tabs` being a `bool` whose `Default` implementation returns `false`).

> Is it necessary to have constants for the boolean defaults?

If they're all `false` by default, then possibly no in favour of using `Default::default()`, however if any are `true` then they should have constants.

> Should I make every default a constant?

Up to you - I think I probably would just for consistency.

Reply:
Thanks so much for the input, other than some warnings for unused methods or constants, this PR should be ready for review tomorrow I think!